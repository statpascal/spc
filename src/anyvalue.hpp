/** \file anyvalue.hpp
*/

#pragma once

#include <algorithm>
#include <iostream>

namespace statpascal {

class TAnyValue {
public:
    TAnyValue ();
    TAnyValue (const TAnyValue &);
    TAnyValue (TAnyValue &&);
    TAnyValue (void *p);
    template<typename T, 
             typename U = typename std::remove_cv<typename std::remove_reference<T>::type>::type,
             typename V = typename std::enable_if<!std::is_same<TAnyValue, U>::value>::type> TAnyValue (T &&);
             
    TAnyValue &operator = (TAnyValue);
    
    ~TAnyValue ();
    
    TAnyValue &swap (TAnyValue &);
    
    /** creates copy if value is shared. */
    void copyOnWrite ();
    
    /** returns true if a value has been stored. */
    bool hasValue () const;
    
    template<typename T> T &get ();
    template<typename T> const T &get () const;
    
private:
    class TValue {
    public:
        TValue (): refcount (1) {};
        virtual ~TValue () = default;
        virtual TValue *clone () = 0;
        
        std::size_t refcount;
    };
    
    template<typename T> class TConcreteValue: public TValue {
    public:
        TConcreteValue (const T &);
        TConcreteValue (T &&);
        virtual ~TConcreteValue () = default;
        
        virtual TValue *clone () override;

        T concreteValue;
    };
    
    TValue *value;
};

template<typename T> inline TAnyValue::TConcreteValue<T>::TConcreteValue (const T &val): concreteValue (val) {
}

template<typename T> inline TAnyValue::TConcreteValue<T>::TConcreteValue (T &&val): concreteValue (std::move (val)) {
}

template<typename T> inline TAnyValue::TValue *TAnyValue::TConcreteValue<T>::clone () {
    return new TConcreteValue (concreteValue);
}

inline TAnyValue::TAnyValue (): 
  value (nullptr) {
}

inline TAnyValue::TAnyValue (void *p):
  value (static_cast<TValue *> (p)) {
      if (value)
          ++value->refcount;
}

inline TAnyValue::~TAnyValue () {
    if (value && !--value->refcount)
        delete value;
}

inline TAnyValue::TAnyValue (const TAnyValue &other):
  value (other.value) {
    if (value)
      ++value->refcount;

}

inline TAnyValue::TAnyValue (TAnyValue &&other):
  value (other.value) {
    other.value = nullptr;
}

template<typename T, typename U, typename V> inline TAnyValue::TAnyValue (T &&val):
  value (new TConcreteValue<U> (std::forward<T> (val))) {
}

inline TAnyValue &TAnyValue::operator = (TAnyValue other) {
    return swap (other);
}

inline TAnyValue &TAnyValue::swap (TAnyValue &other) {
    std::swap (value, other.value);
    return *this;
}

template<typename T> inline T &TAnyValue::get () {
    return static_cast<TConcreteValue<T> *> (value)->concreteValue;
}

template<typename T> inline const T &TAnyValue::get () const {
    return static_cast<const TConcreteValue<T> *> (value)->concreteValue;
}

inline void TAnyValue::copyOnWrite () {
    if (value && value->refcount > 1) {
        --value->refcount;
        value = value->clone ();
    }
}

inline bool TAnyValue::hasValue () const {
    return !!value;
}

}

#ifdef TEST

// Test

#include <iostream>

using namespace statpascal;

class TTestClass {
public:
    TTestClass (int n): n (n) { std::cout << "Konstruktor: " << n << std::endl; }
    ~TTestClass () { std::cout << "Destruktor: " << n << std::endl; }
    TTestClass (const TTestClass &other): n (other.n) { std::cout << "Copy Constructor: " << n << std::endl; }
    TTestClass (TTestClass &&other): n (other.n)  { std::cout << "Move Contructor: " << n << std::endl; other.n = 0; }
    
    int getn () { return n;}
private:
    int n;
};

int main () {
//    TAnyValue any (TTestClass (5));
    
    TTestClass t (7);
    TAnyValue any (TTestClass (6));
    std::cout << "Size is: " << sizeof (any) << std::endl << " " << t.getn () << std::endl;
    TAnyValue any1 (std::move (any));
    TAnyValue any2 = any1;
    TAnyValue any3 = any1;
    any1 = any1;
    std::cout << "Any is: " << any1.get<TTestClass> ().getn () << std::endl;
}

#endif
