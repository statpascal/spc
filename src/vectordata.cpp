#include "vectordata.hpp"
#include "anymanager.hpp"

namespace statpascal {

TVectorDataPtr &TVectorDataPtr::operator = (TVectorDataPtr other) {
    return swap (other);
}

TVectorDataPtr &TVectorDataPtr::swap (TVectorDataPtr &other) {
    std::swap (vectorData, other.vectorData);
    return *this;
}

void TVectorDataPtr::copyOnWrite () {
    if (vectorData && vectorData->refcount > 1) {
        --vectorData->refcount;
        TVectorData *copy = static_cast<TVectorData *> (malloc (sizeof (TVectorData) + vectorData->size * vectorData->count));
        new (copy) TVectorData (*vectorData);
        vectorData = copy;
    }
}

TVectorData::TVectorData (const TVectorData &other):
  size (other.size), count (other.count), refcount (1), anyManager (other.anyManager) {
    if (anyManager)
        for (std::size_t i = 0; i < count; ++i)
            setElement (i, &other.get<unsigned char> (i));
    else
        memcpy (data, other.data, size * count);
}

void TVectorData::setElement (std::size_t index, const void *src) {
    unsigned char *dest = &data [index * size];
    std::memcpy (dest, src, size);
    if (anyManager)
        anyManager->copy (src, dest);
}

void TVectorData::deleteData () {
    if (anyManager)        
        for (std::size_t i = 0; i < count; ++i)
            anyManager->destroy (&get<unsigned char> (i));
}

}