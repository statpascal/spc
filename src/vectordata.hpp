/** \file vectordata.hpp
*/

#pragma once

#include <cstddef>
#include <cstring>
#include <cstdlib>
#include <new>
#include <algorithm>

namespace statpascal {

class TVectorData;
class TAnyManager;

class TVectorDataPtr {
public:
    TVectorDataPtr ();
    TVectorDataPtr (const TVectorDataPtr &);
    explicit TVectorDataPtr (void *, bool incRef = true);
    TVectorDataPtr (std::size_t elementSize, std::size_t elementCount, TAnyManager *elementAnyManager = nullptr, bool zeroMemory = true);
    
    TVectorDataPtr &operator = (TVectorDataPtr);
    
    ~TVectorDataPtr ();

    TVectorDataPtr &swap (TVectorDataPtr &);
    
    void copyOnWrite ();
    bool hasValue () const;
    
    TVectorData &operator * () const;
    TVectorData *operator -> () const;
    
    TVectorData *makeRef () const;
    void releaseRef ();
    
private:
    TVectorData *vectorData;
};


class TVectorData {
public:
    void setElement (std::size_t index, const void *src);
    const void *getElement (std::size_t index) const;
    void *getElement (std::size_t index);
    
    template<typename T> T &get (std::size_t index);
    template<typename T> const T &get (std::size_t index) const;
    
    TAnyManager *getElementAnyManager () const;
    std::size_t getElementSize () const;
    std::size_t getElementCount () const;
    std::size_t getRefCount () const;

private:
    TVectorData (std::size_t elementSize, std::size_t elementCount, TAnyManager *elementAnyManager = nullptr, bool zeroMemory = true);
    TVectorData (const TVectorData &);
    ~TVectorData ();
    
    TVectorData &operator = (TVectorData) = delete;
    
    void deleteData ();

    std::size_t size, count, refcount;
    TAnyManager *anyManager;
    unsigned char data [1];
    
    friend TVectorDataPtr;
};


inline TVectorData::TVectorData (std::size_t size, std::size_t count, TAnyManager *anyManager, bool zeroMemory):
  size (size), count (count), refcount (1), anyManager (anyManager) {
    if (zeroMemory)
        std::fill (data, data + size * count, 0);
}

inline TVectorDataPtr::TVectorDataPtr (std::size_t elementSize, std::size_t elementCount, TAnyManager *elementAnyManager, bool zeroMemory):
  vectorData (static_cast<TVectorData *> (malloc (sizeof (TVectorData) + elementSize * elementCount))) {
    new (vectorData) TVectorData (elementSize, elementCount, elementAnyManager, zeroMemory);
}

inline TVectorData *TVectorDataPtr::operator -> () const {
    return vectorData;
}

inline TVectorData *TVectorDataPtr::makeRef () const {
    if (vectorData)
        ++vectorData->refcount;
    return vectorData;
}

inline void TVectorDataPtr::releaseRef () {
    if (vectorData && !--vectorData->refcount) {
        vectorData->~TVectorData ();
        free (vectorData);
    }
}

inline TVectorData::~TVectorData () {
    if (anyManager)
        deleteData ();
}

inline TAnyManager *TVectorData::getElementAnyManager () const {
    return anyManager;
}

inline std::size_t TVectorData::getElementSize () const {
    return size;
}

inline std::size_t TVectorData::getRefCount () const {
    return refcount;
}

inline std::size_t TVectorData::getElementCount () const {
    return count;
}

inline const void *TVectorData::getElement (std::size_t index) const {
    return &data [index * size];
}

inline void *TVectorData::getElement (std::size_t index) {
    return &data [index * size];
}

template<typename T> inline const T &TVectorData::get (std::size_t index) const {
    return *static_cast<const T *> (getElement (index));
}

template<typename T> inline T &TVectorData::get (std::size_t index) {
    return const_cast<T &> (static_cast<const TVectorData *> (this)->get<T> (index));
}

inline TVectorDataPtr::TVectorDataPtr ():
  vectorData (nullptr) {
}

inline TVectorDataPtr::TVectorDataPtr (const TVectorDataPtr &other):
  vectorData (other.makeRef ()) {
}

inline TVectorDataPtr::TVectorDataPtr (void *p, bool incRef):
  vectorData (static_cast<TVectorData *> (p)) {
    if (incRef && vectorData)
        ++vectorData->refcount;
}

inline TVectorDataPtr::~TVectorDataPtr () {
    releaseRef ();
}

inline bool TVectorDataPtr::hasValue () const {
    return !!vectorData;
}

inline TVectorData &TVectorDataPtr::operator * () const {
    return *vectorData;
}

}
