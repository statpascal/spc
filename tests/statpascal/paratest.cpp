#include <stdio.h>
#include <string>
#include <iostream>
#include <cmath>

#include "anyvalue.hpp"

#include <stdio.h>
#include <string>
#include <iostream>
#include <cmath>

#include "anyvalue.hpp"
#include "vectordata.hpp"

extern "C" statpascal::TAnyValue manyparastr (statpascal::TAnyValue a, statpascal::TAnyValue b, statpascal::TAnyValue c, statpascal::TAnyValue d, statpascal::TAnyValue e, statpascal::TAnyValue f, statpascal::TAnyValue g, statpascal::TAnyValue h, statpascal::TAnyValue i, statpascal::TAnyValue j) {
    return h;
}

extern "C" int manypara (int a, int b, int c, int d, int e, int f, int g, int h, int i, int j) {
    return i + j;
}

