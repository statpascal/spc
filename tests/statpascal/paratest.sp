program paratest;

const
    libname = 'paratest.so';
    n = 100;

type 
    rec = record
	a: int64;
	s1: string;
	b: real;
	s2: string;
	c: char
    end;
    arr = array [1..n] of integer;

function manypara (a, b, c, d, e, f, g, h, i, j: integer): integer; external libname;
function manyparastr (a, b, c, d, e, f, g, h, i, j: string): string; external libname;

function manypara1 (a, b, c, d, e, f, g, h, i, j: integer): integer;
    begin
	writeln (a, b, c, d, e, f, g, h, i, j);
	manypara1 := h
    end;

function manystr (a, b, c, d, e, f, g, h, i, j: string): string;
    begin
	writeln (a, b, c, d, e, f, g, h, i, j);
	manystr := i
    end;

function manyrec (a, b, c, d, e, f, g, h, i, j: rec): rec;
    begin
	manyrec := j
    end;

function manyarr (a, b, c, d, e, f, g, h, i, j: arr): arr;
    begin
	manyarr := h
    end;

procedure dotest;
    var
        a: integer;
        s: string;
	i: 0..9;
	j: 1..n;
	r: array [0..9] of rec;
	r1: rec;
	f: array [0..9] of arr;
	f1: arr;

    function func (i: integer): rec;
	begin
	    func := r [i]
	end;

    begin
        a := manypara (1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
        writeln (a);
        a := manypara1 (1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
        writeln (a);
        s := manystr ('S1', 'S2', 'S3', 'S4', 'S5', 'S6', 'S7', 'S8', 'S9', 'S10');
        writeln (s);
        s := manyparastr ('S1', 'S2', 'S3', 'S4', 'S5', 'S6', 'S7', 'S8', 'S9', 'S10');
        writeln (s);
	for i := 0 to 9 do
	    with r [i] do
		begin
		    a := i;
		    str (i, s1);
		    b := i + 0.1;
		    str (i, s2);
		    c := chr (i + 65)
		end;
	r1 := manyrec (r [0], r [1], r [2], r [3], r [4], r [5], r [6], r [7], r [8], r [9]);
	writeln (r1.s2);
	r1 := manyrec (func (0), func (1), func (2), func (3), func (4), func (5), func (6), func (7), func (8), func (9));
	writeln (r1.s2);
	for i := 0 to 9 do
	    for j := 1 to n do
		f [i, j] := i * n + j;
	f1 := manyarr (f [0], f [1], f [2], f [3], f [4], f [5], f [6], f [7], f [8], f [9]);
	writeln (f1 [50]);
	writeln (manyarr (f [0], f [1], f [2], f [3], f [4], f [5], f [6], f [7], f [8], f [9]) [34])
    end;

begin
    dotest
end.
