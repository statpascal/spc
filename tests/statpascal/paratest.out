19
12345678910
8
S1S2S3S4S5S6S7S8S9S10
S9
S8
9
9
750
734
